File: solutions/Medium/Permutations_0046.js
/**
 * Function: permute
 * Description: Given an array of distinct integers `nums`, this function returns all possible 
 * permutations of the elements.
 *
 * @param {number[]} nums - An array of distinct integers.
 * @returns {number[][]} - A list containing all permutations of `nums`.
 */
const permute = function(nums) {
    let result = [];

    /**
     * Backtracking function to generate permutations.
     * @param {number[]} current - The current permutation being formed.
     * @param {boolean[]} used - Boolean array to track used elements.
     */
    const backtrack = (current, used) => {
        // If current permutation has all numbers, add it to result
        if (current.length === nums.length) {
            result.push([...current]);
            return;
        }

        // Iterate through nums and construct permutations
        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue; // Skip already used numbers

            used[i] = true; // Mark current number as used
            current.push(nums[i]); // Add number to the current permutation
            backtrack(current, used); // Recurse with updated state
            current.pop(); // Backtrack: remove last number
            used[i] = false; // Reset used flag
        }
    };

    // Start backtracking with an empty permutation
    backtrack([], new Array(nums.length).fill(false));

    return result;
};

// Example Test Cases
if (require.main === module) {
    console.log("Test 1: Basic case with three numbers");
    console.log(permute([1, 2, 3]));
    // Expected output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

    console.log("Test 2: Case with two numbers");
    console.log(permute([0, 1]));
    // Expected output: [[0,1],[1,0]]

    console.log("Test 3: Single number case");
    console.log(permute([1]));
    // Expected output: [[1]]

    console.log("Test 4: Case with four numbers");
    console.log(permute([1, 2, 3, 4]));
    // Expected output: 24 unique permutations

    console.log("Test 5: Case with negative numbers");
    console.log(permute([-1, 2, -3]));
    // Expected output: All possible orders of [-1,2,-3]

    console.log("Test 6: Case with the largest input (6 numbers)");
    console.log(permute([1, 2, 3, 4, 5, 6]).length);
    // Expected output: 6! = 720
}

/* Complexity analysis:
Time Complexity: O(n!) - There are n! permutations for an array of size n.
Space Complexity: O(n!) - Due to storing all permutations in the result array.
*/

export default permute;

File: solutions/Medium/Permutations_0046.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        """
        Function: permute
        Description: Generates all possible permutations of a given list of distinct integers.

        Parameters:
        - nums (List[int]): List of distinct integers.

        Returns:
        - List[List[int]]: A list of all possible permutations.
        """

        res = []

        def backtrack(path, available):
            # Base case: If the path contains all elements, store the permutation
            if not available:
                res.append(path[:])  # Make a copy before adding
                return

            # Iterate through remaining elements
            for i in range(len(available)):
                backtrack(path + [available[i]], available[:i] +
                          available[i+1:])  # Choose and explore

        backtrack([], nums)

        return res


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Standard case with three elements
    print("Test Case 1: nums = [1,2,3]")
    print("Expected Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]] | Actual Output:", solution.permute(
        [1, 2, 3]))

    # Test case 2: Two elements (simplest multi-permutation case)
    print("Test Case 2: nums = [0,1]")
    print("Expected Output: [[0,1],[1,0]] | Actual Output:",
          solution.permute([0, 1]))

    # Test case 3: Single element
    print("Test Case 3: nums = [1]")
    print("Expected Output: [[1]] | Actual Output:", solution.permute([1]))

    # Test case 4: Largest input size (6 elements)
    print("Test Case 4: nums = [1,2,3,4,5,6]")
    print("Expected Output: Computed Output:",
          solution.permute([1, 2, 3, 4, 5, 6]))

    # Test case 5: Permutations of negative numbers
    print("Test Case 5: nums = [-1, -2, -3]")
    print("Expected Output: Computed Output:", solution.permute([-1, -2, -3]))

    # Test case 6: Mixed positive and negative numbers
    print("Test Case 6: nums = [-1, 0, 1]")
    print("Expected Output: Computed Output:", solution.permute([-1, 0, 1]))

"""
Time Complexity:
- O(N!) → There are N! permutations to generate.
- Space Complexity:
  - O(N) → Recursion depth is at most N.
"""
